link-grammar/connectors.c

```diff
@@ -33,116 +35,221 @@ void free_connectors(Connector *e)
 	}
 }
 
-Connector * connector_new(void)
+void
+set_connector_length_limit(Connector *c, Parse_Options opts)
+{
+	if (NULL == opts)
+	{
+		c->length_limit = UNLIMITED_LEN;
+		return;
+	}
+
+	int short_len = opts->short_length;
+	bool all_short = opts->all_short;
+	int length_limit = c->desc->length_limit;
+
+	if ((all_short && (length_limit > short_len)) || (0 == length_limit))
+		c->length_limit = short_len;
+	else
+		c->length_limit = length_limit;
+}
+
+Connector * connector_new(const condesc_t *desc, Parse_Options opts)
 {
 	Connector *c = (Connector *) xalloc(sizeof(Connector));
-	init_connector(c);
+
+	c->desc = desc;
 	c->nearest_word = 0;
 	c->multi = false;
-	c->lc_start = 0;
-	c->uc_length = 0;
-	c->uc_start = 0;
-	c->next = NULL;
-	c->string = "";
-	c->tableNext = NULL;
+	set_connector_length_limit(c, opts);
+	//assert(0 != c->length_limit, "Connector_new(): Zero length_limit");
+
 	return c;
 }
 
 /* ======================================================== */
-/* Connector-set utilities ... */
-/**
- * This hash function only looks at the leading upper case letters of
- * the string, and the direction, '+' or '-'.
- */
-static unsigned int connector_set_hash(Connector_set *conset, const char * s, int d)
-{
-	unsigned int i;
-	if (islower((int)*s)) s++; /* skip head-dependent indicator */
-
-	/* djb2 hash */
-	i = 5381;
-	i = ((i << 5) + i) + d;
-	while (isupper((int) *s)) /* connector tables cannot contain UTF8, yet */
-	{
-		i = ((i << 5) + i) + *s;
-		s++;
-	}
-	return (i & (conset->table_size-1));
-}
+/* UNLIMITED-CONNECTORS handling. */
 
-static void build_connector_set_from_expression(Connector_set * conset, Exp * e)
+static void get_connectors_from_expression(condesc_t **conlist, size_t *cl_size,
+                                           const Exp *e)
 {
 	E_list * l;
-	Connector * c;
-	unsigned int h;
+
 	if (e->type == CONNECTOR_type)
 	{
-		c = connector_new();
-		c->string = e->u.string;
-		h = connector_set_hash(conset, c->string, e->dir);
-		c->next = conset->hash_table[h];
-		conset->hash_table[h] = c;
+		if (NULL != conlist)
+		{
+			conlist[*cl_size] = e->u.condesc;
+		}
+		(*cl_size)++;
 	} else {
 		for (l=e->u.l; l!=NULL; l=l->next) {
-			build_connector_set_from_expression(conset, l->e);
+			get_connectors_from_expression(conlist, cl_size, l->e);
 		}
```

Comment:
I think it would be slightly faster to return cl_size as a return value, rather than passing a pointer.  When you increment a value at a location, the compiler has to assume a worst-case of possible aliasing, and constantly write it out to memory (cache). By contrast, a return value will usually stay in registers, maybe be written to stack occasionally.

Reply:
You are right. I will fix that.